[Bug] Umbrella package was corrupted with upToNextMajorVersion/upToNextMinorVersion

Hey @trinhngocthuyen 

Tried this yesterday but got a bunch of errors, the packages weren't being correctly setup in xcode.

The error logs look like this:

```
Resolving SPM dependencies
error: '.umbrella': Invalid manifest (compiled with: ["/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc", "-vfsoverlay", "/var/folders/1w/l2ht_k557nz44xqkw21zbd7w0000gp/T/TemporaryDirectory.imvefZ/vfs.yaml", "-L", "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI", "-lPackageDescription", "-Xlinker", "-rpath", "-Xlinker", "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI", "-target", "arm64-apple-macosx13.0", "-sdk", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk", "-F", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks", "-I", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib", "-L", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib", "-swift-version", "5", "-I", "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/pm/ManifestAPI", "-sdk", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk", "-package-description-version", "5.7.0", "/Users/me/project/.spm.pods/packages/.umbrella/Package.swift", "-o", "/var/folders/1w/l2ht_k557nz44xqkw21zbd7w0000gp/T/TemporaryDirectory.KRiGJ4/.umbrella-manifest"])
/Users/me/project/.spm.pods/packages/.umbrella/Package.swift:26:6: error: no exact matches in call to static method 'package'
24 |     .package(url: "a.git.url.1.git", exact: "9.0.0"),
25 |     .package(url: "a.git.url.2.git", exact: "4.1.0"),
26 |     .package(url: "https://github.com/ra1028/DifferenceKit.git", todo: "handle this"),
   |      `- error: no exact matches in call to static method 'package'
27 |     .package(url: "a.git.url.3.git", exact: "3.0.0"),
28 |     .package(path: "a.path"),

PackageDescription.Package.Dependency:6:24: note: incorrect labels for candidate (have: '(url:todo:)', expected: '(url:branch:)')
 4 |     public static func package(name: String, url: String, from version: Version) -> Package.Dependency
 5 |     @available(_PackageDescription 5.5)
 6 |     public static func package(url: String, branch: String) -> Package.Dependency
   |                        `- note: incorrect labels for candidate (have: '(url:todo:)', expected: '(url:branch:)')
 7 |     @available(_PackageDescription, introduced: 5.5, deprecated: 5.6, message: "use package(url:branch:) instead")
 8 |     public static func package(name: String, url: String, branch: String) -> Package.Dependency
 9 |     @available(_PackageDescription 5.5)
10 |     public static func package(url: String, revision: String) -> Package.Dependency
   |                        `- note: incorrect labels for candidate (have: '(url:todo:)', expected: '(url:revision:)')
11 |     @available(_PackageDescription, introduced: 5.5, deprecated: 5.6, message: "use package(url:revision:) instead")
12 |     public static func package(name: String, url: String, revision: String) -> Package.Dependency
[!] Package description of `ABCABC` does not exist!

```

I simply changed from this:

`    spm_pkg 'DifferenceKit', :url => 'https://github.com/ra1028/DifferenceKit.git', :version => '1.3.0'
`
to this:

`    spm_pkg 'DifferenceKit', :url => 'https://github.com/ra1028/DifferenceKit.git', :up_to_next_major_version => '1.3.0'
`

I can try to create a proper issue later if you're unable to reproduce.

_Originally posted by @diogobalseiro in https://github.com/trinhngocthuyen/cocoapods-spm/issues/149#issuecomment-2665089443_

--- a/examples/Podfile
+++ b/examples/Podfile
@@ -49,7 +49,9 @@ target "EX" do # rubocop:disable Metrics/BlockLength
           :version => "5.7.1",
           :products => ["SnapKit-Dynamic"]
   spm_pkg "SwiftUIX", :git => "https://github.com/SwiftUIX/SwiftUIX.git", :tag => "0.1.9"
-  spm_pkg "SwiftyBeaver", :git => "https://github.com/SwiftyBeaver/SwiftyBeaver.git", :tag => "2.0.0"
+  spm_pkg "SwiftyBeaver",
+          :git => "https://github.com/SwiftyBeaver/SwiftyBeaver.git",
+          :up_to_next_major_version => "2.0.0"
   spm_pkg "opentelemetry-swift",
           :git => "https://github.com/open-telemetry/opentelemetry-swift.git",
           :branch => "main",

--- a/lib/cocoapods-spm/resolver/umbrella_package.rb
+++ b/lib/cocoapods-spm/resolver/umbrella_package.rb
@@ -23,17 +23,23 @@ module Pod
           # https://developer.apple.com/documentation/packagedescription/package/dependency
           next ".package(path: \"#{pkg.absolute_path}\")" if pkg.local?
 
-          tail = case pkg.requirement[:kind]
-                 when "exactVersion"
-                   "exact: \"#{pkg.requirement[:version]}\""
-                 when "branch"
-                   "branch: \"#{pkg.requirement[:branch]}\""
-                 when "revision"
-                   "revision: \"#{pkg.requirement[:revision]}\""
-                 else
-                   # TODO: Handle this
-                   "todo: \"handle this\""
-                 end
+          tail =
+            case pkg.requirement[:kind]
+            when "exactVersion"
+              "exact: \"#{pkg.requirement[:version]}\""
+            when "branch"
+              "branch: \"#{pkg.requirement[:branch]}\""
+            when "revision"
+              "revision: \"#{pkg.requirement[:revision]}\""
+            when "upToNextMajorVersion"
+              ".upToNextMajor(from: \"#{pkg.requirement[:minimumVersion]}\")"
+            when "upToNextMinorVersion"
+              ".upToNextMinor(from: \"#{pkg.requirement[:minimumVersion]}\")"
+            when "versionRange"
+              "\"#{pkg.requirement[:minimumVersion]}\"..<\"#{pkg.requirement[:maximumVersion]}\""
+            else
+              raise Informative, "Unexpected pkg requirement: #{pkg.requirement}"
+            end
           ".package(url: \"#{pkg.url}\", #{tail})"
         end
 

